File: examples/cpp/write_modes.cpp
/**
 * @file write_modes.cpp
 * @brief Demonstrates O_DIRECT vs buffered I/O with AuraIO (C++ version)
 *
 * Shows that AuraIO is agnostic to I/O mode - it works with both
 * O_DIRECT and buffered writes. The adaptive tuning measures actual
 * completion latency regardless of how the kernel handles the I/O.
 *
 * Usage: ./write_modes <file>
 */

#include <auraio.hpp>

#include <iostream>
#include <iomanip>
#include <chrono>
#include <vector>
#include <cstring>
#include <fcntl.h>
#include <unistd.h>

using Clock = std::chrono::high_resolution_clock;

constexpr size_t WRITE_SIZE = 64 * 1024;  // 64KB per write
constexpr int NUM_WRITES = 16;            // Number of writes per test

/**
 * Run write test with specified mode.
 */
int run_write_test(const char* filename, bool use_direct) {
    const char* mode_name = use_direct ? "O_DIRECT" : "Buffered";
    std::cout << "\n=== Testing " << mode_name << " writes ===\n";

    try {
        // Create engine
        auraio::Engine engine;

        // Open file with appropriate flags
        int flags = O_WRONLY | O_CREAT | O_TRUNC;
        if (use_direct) {
            flags |= O_DIRECT;
        }

        int fd = open(filename, flags, 0644);
        if (fd < 0) {
            if (use_direct && errno == EINVAL) {
                std::cout << "O_DIRECT not supported on this filesystem, skipping\n";
                return 0;
            }
            throw auraio::Error(errno, filename);
        }

        // Allocate buffer - use aligned for O_DIRECT, regular vector for buffered
        auraio::Buffer aligned_buf;
        std::vector<char> heap_buf;
        void* buf;

        if (use_direct) {
            aligned_buf = engine.allocate_buffer(WRITE_SIZE);
            buf = aligned_buf.data();
            std::cout << "Using aligned buffer from engine.allocate_buffer()\n";
        } else {
            heap_buf.resize(WRITE_SIZE);
            buf = heap_buf.data();
            std::cout << "Using std::vector buffer\n";
        }

        // Fill buffer with pattern
        std::memset(buf, 'A', WRITE_SIZE);

        // Track completion
        int completed = 0;
        auto start_time = Clock::now();
        Clock::time_point end_time;

        // Submit all writes
        std::cout << "Submitting " << NUM_WRITES << " async writes of "
                  << WRITE_SIZE << " bytes each...\n";

        for (int i = 0; i < NUM_WRITES; i++) {
            off_t offset = i * static_cast<off_t>(WRITE_SIZE);
            engine.write(fd, auraio::buf(buf), WRITE_SIZE, offset,
                [&completed, &end_time](auraio::Request&, ssize_t result) {
                    if (result < 0) {
                        std::cerr << "Write failed: " << strerror(static_cast<int>(-result)) << "\n";
                    }
                    completed++;
                    if (completed == NUM_WRITES) {
                        end_time = Clock::now();
                    }
                });
        }

        // Wait for all completions
        while (completed < NUM_WRITES) {
            engine.wait(100);
        }

        // Calculate results
        double elapsed_sec = std::chrono::duration<double>(end_time - start_time).count();
        double total_mb = static_cast<double>(NUM_WRITES * WRITE_SIZE) / (1024 * 1024);
        double throughput = total_mb / elapsed_sec;

        std::cout << "Completed " << completed << " writes in "
                  << std::fixed << std::setprecision(3) << elapsed_sec << " seconds\n";
        std::cout << "Total: " << std::setprecision(2) << total_mb
                  << " MB, Throughput: " << throughput << " MB/s\n";

        // Get engine stats
        auto stats = engine.get_stats();
        std::cout << "P99 latency: " << stats.p99_latency_ms() << " ms\n";
        std::cout << "Optimal in-flight: " << stats.optimal_in_flight() << "\n";

        // Fsync to ensure data is on disk
        std::cout << "Fsyncing...\n";
        fsync(fd);

        // Cleanup (buffers cleaned up by RAII)
        close(fd);

        return 0;

    } catch (const auraio::Error& e) {
        std::cerr << "AuraIO error: " << e.what() << "\n";
        return -1;
    }
}

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <file>\n";
        std::cerr << "\nDemonstrates O_DIRECT vs buffered async writes.\n";
        std::cerr << "The file will be created/overwritten.\n";
        return 1;
    }

    const char* filename = argv[1];

    std::cout << "AuraIO Write Modes Example (C++)\n";
    std::cout << "================================\n";
    std::cout << "File: " << filename << "\n";
    std::cout << "Write size: " << WRITE_SIZE << " bytes\n";
    std::cout << "Number of writes: " << NUM_WRITES << "\n";

    // Test buffered I/O first
    if (run_write_test(filename, false) < 0) {
        return 1;
    }

    // Test O_DIRECT
    if (run_write_test(filename, true) < 0) {
        return 1;
    }

    // Cleanup test file
    unlink(filename);

    std::cout << "\n=== Summary ===\n";
    std::cout << "Both modes use the same AuraIO API.\n";
    std::cout << "The library is agnostic - it just submits to io_uring.\n";
    std::cout << "O_DIRECT: Requires aligned buffers, bypasses page cache.\n";
    std::cout << "Buffered: Any buffer works, uses page cache.\n";

    return 0;
}

File: examples/rust/examples/write_modes.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 AuraIO Contributors


//! Demonstrates O_DIRECT vs buffered I/O with AuraIO
//!
//! Shows that AuraIO is agnostic to I/O mode - it works with both
//! O_DIRECT and buffered writes. The adaptive tuning measures actual
//! completion latency regardless of how the kernel handles the I/O.
//!
//! Usage: cargo run --example write_modes -- <file>

use aura::{Engine, Result};
use std::env;
use std::fs;
use std::os::unix::io::RawFd;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use std::time::Instant;

const WRITE_SIZE: usize = 64 * 1024; // 64KB per write
const NUM_WRITES: usize = 16; // Number of writes per test

/// Completion tracking
struct WriteState {
    completed: AtomicUsize,
    total: usize,
    start_time: Instant,
}

/// Run write test with specified mode.
fn run_write_test(filename: &str, use_direct: bool) -> Result<()> {
    let mode_name = if use_direct { "O_DIRECT" } else { "Buffered" };
    println!("\n=== Testing {} writes ===", mode_name);

    // Create engine
    let engine = Engine::new()?;

    // Open file with appropriate flags
    let mut flags = libc::O_WRONLY | libc::O_CREAT | libc::O_TRUNC;
    if use_direct {
        flags |= libc::O_DIRECT;
    }

    let fd: RawFd = unsafe {
        let c_path = std::ffi::CString::new(filename).unwrap();
        libc::open(c_path.as_ptr(), flags, 0o644)
    };

    if fd < 0 {
        let err = std::io::Error::last_os_error();
        if use_direct && err.raw_os_error() == Some(libc::EINVAL) {
            println!("O_DIRECT not supported on this filesystem, skipping");
            return Ok(());
        }
        eprintln!("Failed to open '{}': {}", filename, err);
        return Err(aura::Error::Io(err));
    }

    // Allocate buffer - use aligned for O_DIRECT
    let buf = if use_direct {
        println!("Using aligned buffer from engine.allocate_buffer()");
        let b = engine.allocate_buffer(WRITE_SIZE)?;
        // Fill with pattern
        // Note: we can't easily modify the buffer after allocation in this API
        b
    } else {
        println!("Using aligned buffer (Rust version always uses engine pool)");
        let b = engine.allocate_buffer(WRITE_SIZE)?;
        b
    };

    // Track completion
    let state = Arc::new(WriteState {
        completed: AtomicUsize::new(0),
        total: NUM_WRITES,
        start_time: Instant::now(),
    });

    // Submit all writes
    println!("Submitting {} async writes of {} bytes each...", NUM_WRITES, WRITE_SIZE);

    for i in 0..NUM_WRITES {
        let offset = (i * WRITE_SIZE) as i64;
        let state_clone = state.clone();

        unsafe {
            engine.write(fd, (&buf).into(), WRITE_SIZE, offset, move |result| {
                if let Err(e) = result {
                    eprintln!("Write failed: {}", e);
                }
                state_clone.completed.fetch_add(1, Ordering::SeqCst);
            })?;
        }
    }

    // Wait for all completions
    while state.completed.load(Ordering::SeqCst) < state.total {
        engine.wait(100)?;
    }

    // Calculate results
    let elapsed_sec = state.start_time.elapsed().as_secs_f64();
    let total_mb = (NUM_WRITES * WRITE_SIZE) as f64 / (1024.0 * 1024.0);
    let throughput = total_mb / elapsed_sec;

    println!("Completed {} writes in {:.3} seconds", state.total, elapsed_sec);
    println!("Total: {:.2} MB, Throughput: {:.2} MB/s", total_mb, throughput);

    // Get engine stats
    let stats = engine.stats()?;
    println!("P99 latency: {:.2} ms", stats.p99_latency_ms());
    println!("Optimal in-flight: {}", stats.optimal_in_flight());

    // Fsync to ensure data is on disk
    println!("Fsyncing...");
    unsafe { libc::fsync(fd); }

    // Cleanup
    unsafe { libc::close(fd); }

    Ok(())
}

fn main() -> Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <file>", args[0]);
        eprintln!("\nDemonstrates O_DIRECT vs buffered async writes.");
        eprintln!("The file will be created/overwritten.");
        std::process::exit(1);
    }

    let filename = &args[1];

    println!("AuraIO Write Modes Example (Rust)");
    println!("==================================");
    println!("File: {}", filename);
    println!("Write size: {} bytes", WRITE_SIZE);
    println!("Number of writes: {}", NUM_WRITES);

    // Test buffered I/O first
    run_write_test(filename, false)?;

    // Test O_DIRECT
    run_write_test(filename, true)?;

    // Cleanup test file
    fs::remove_file(filename).ok();

    println!("\n=== Summary ===");
    println!("Both modes use the same AuraIO API.");
    println!("The library is agnostic - it just submits to io_uring.");
    println!("O_DIRECT: Requires aligned buffers, bypasses page cache.");
    println!("Buffered: Any buffer works, uses page cache.");

    Ok(())
}
